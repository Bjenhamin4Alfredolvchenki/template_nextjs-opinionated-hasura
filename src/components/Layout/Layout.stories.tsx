import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'
import { Layout, LayoutProps } from './Layout'
import { LinksList } from '../../model/site/LinksList'

export default {
  title: 'Component/Layout',
  component: Layout,
  argTypes: {},
} as Meta

const Template: Story<LayoutProps> = (args) => (
  <div>
    <Layout {...args} />
  </div>
)

// Default scenario
export const MainLayout = Template.bind({})
MainLayout.args = {
  title: (
    <div className='flex items-baseline flex-grow px-2 mx-2 space-x-3'>
      <div className='text-base font-bold'>HOME</div>
      <div className='text-sm'>Next.js Opinionated</div>
    </div>
  ),
  menuItems: Object.values(LinksList),
  children:
    'Lorem ipsum dolor sit amet, consectetur adipiscing. Vestibulum rutrum metus at enim congue scelerisque. Sed suscipit metu non iaculis semper consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing. Vestibulum rutrum metus at enim congue scelerisque. Sed suscipit metu non iaculis semper consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing. Vestibulum rutrum metus at enim congue scelerisque. Sed suscipit metu non iaculis semper consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing. Vestibulum rutrum metus at enim congue scelerisque. Sed suscipit metu non iaculis semper consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing. Vestibulum rutrum metus at enim congue scelerisque. Sed suscipit metu non iaculis semper consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing. Vestibulum rutrum metus at enim congue scelerisque. Sed suscipit metu non iaculis semper consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing. Vestibulum rutrum metus at enim congue scelerisque. Sed suscipit metu non iaculis semper consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing. Vestibulum rutrum metus at enim congue scelerisque. Sed suscipit metu non iaculis semper consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing. Vestibulum rutrum metus at enim congue scelerisque. Sed suscipit metu non iaculis semper consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing. Vestibulum rutrum metus at enim congue scelerisque. Sed suscipit metu non iaculis semper consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing. Vestibulum rutrum metus at enim congue scelerisque. Sed suscipit metu non iaculis semper consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing. Vestibulum rutrum metus at enim congue scelerisque. Sed suscipit metu non iaculis semper consectetur adipiscing elit.',
}
