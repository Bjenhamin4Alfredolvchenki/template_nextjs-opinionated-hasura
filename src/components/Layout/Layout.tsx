import React, { ReactNode, useRef } from 'react'
import { useRouter } from 'next/router'
import Link from 'next/link'
import classnames from 'classnames'
import { LinkProps } from '../../model/site/LinksList'
import { ThemeList } from '../../model/site/ThemeList'
import { useTheme } from 'next-themes'
import { useUser } from '@auth0/nextjs-auth0'
import Loading from '../Loading/Loading'
import { BiLinkExternal, BiLogIn, BiLogOut } from 'react-icons/bi'
import { BsPersonFill } from 'react-icons/bs'
import { DropDown } from '../DropDown/DropDown'
import _ from 'lodash'

export interface LayoutProps {
  title?: ReactNode
  menuItems?: LinkProps[]
  children?: ReactNode
  internal_theme?: string
}

export const Layout: React.FC<LayoutProps> = ({ title, menuItems, children }) => {
  const { theme, setTheme } = useTheme()
  const router = useRouter()
  const { user, error, isLoading } = useUser()
  const checkboxRef = useRef<HTMLInputElement>(null)

  if (isLoading) return <Loading />
  if (error) return <div>{error.message}</div>

  return (
    <div className='h-screen bg-base-100 drawer text-base-content'>
      {/* put everything in a off-canvas drawer */}
      {/* this checkbox controls if drawer is open */}
      <input id='menu-drawer' type='checkbox' className='drawer-toggle' ref={checkboxRef} />
      <div className='flex flex-col drawer-content'>
        {/* drawer content */}
        <div
          className={classnames(
            'inset-x-0 top-0 z-10 w-full border-b border-transparent navbar text-base-content',
            {
              'bg-transparent fixed text-primary-content': router.asPath === '/',
            }
          )}
        >
          {/* hamburger menu is only visible on mobile */}
          <div className='flex-none lg:hidden'>
            <label htmlFor='menu-drawer' className='btn btn-square btn-ghost'>
              <svg
                xmlns='http://www.w3.org/2000/svg'
                fill='none'
                viewBox='0 0 24 24'
                className='inline-block w-6 h-6 stroke-current'
              >
                <path
                  strokeLinecap='round'
                  strokeLinejoin='round'
                  strokeWidth={2}
                  d='M4 6h16M4 12h16M4 18h16'
                />
              </svg>
            </label>
          </div>

          {/* title */}
          {title}

          <div className='flex-none'>
            {/* navbar is only visible for desktop */}
            <div className='hidden lg:inline-block'>
              <ul className='space-x-2 menu horizontal'>
                {menuItems?.map((m) => (
                  <div className='flex items-center' key={m.name}>
                    {m.internalURL && (
                      <Link href={m.internalURL}>
                        <a className='rounded-btn btn btn-ghost' rel='noopener noreferrer'>
                          {m.name}
                        </a>
                      </Link>
                    )}
                    {m.externalURL && (
                      <a
                        className='rounded-btn btn btn-ghost'
                        href={m.externalURL}
                        target='_blank'
                        rel='noopener noreferrer'
                      >
                        {m.name}
                        <BiLinkExternal className='ml-2' size={17} />
                      </a>
                    )}
                  </div>
                ))}
                <li>
                  <div className='z-30 m-1'>
                    <select
                      className='w-full max-w-xs select select-bordered bg-base-100 text-base-content'
                      onChange={(ev) => {
                        setTheme(ev.target.value)
                      }}
                      value={theme}
                    >
                      <option disabled>theme</option>
                      {_.map(ThemeList, (t) => (
                        <option key={t.id} value={t.id}>
                          {t.name}
                        </option>
                      ))}
                    </select>
                  </div>
                </li>
                <li>
                  <div className='m-1'>
                    {!user && (
                      <Link href='/api/auth/login'>
                        <a className='btn btn-ghost'>
                          Login
                          <BiLogIn size={20} className='mx-2' />
                        </a>
                      </Link>
                    )}

                    {user && (
                      <>
                        <DropDown
                          className='static bg-transparent'
                          width={50}
                          selectedId={null}
                          onSelect={() => {
                            /**/
                          }}
                          label={
                            <div className='avatar'>
                              <div className='w-10 h-10 m-1 rounded-full'>
                                {user?.picture ? (
                                  <img alt={user.name || undefined} src={user?.picture} />
                                ) : (
                                  <BsPersonFill />
                                )}
                              </div>
                            </div>
                          }
                          items={[
                            {
                              id: '1',
                              value: (
                                <Link href='/api/auth/logout'>
                                  <a className='justify-center btn btn-ghost'>
                                    <BiLogOut size={20} className='mx-2' />
                                  </a>
                                </Link>
                              ),
                            },
                          ]}
                        />
                      </>
                    )}
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </div>

        {/* main content */}
        <div
          className={classnames('w-full', {
            'p-4 lg:px-6 lg:container lg:mx-auto': router.asPath !== '/',
          })}
        >
          {children}
        </div>
      </div>

      {/* drawer sidebar for mobile */}
      <div className='drawer-side'>
        <label htmlFor='menu-drawer' className='drawer-overlay' />
        <ul className='p-4 overflow-y-auto menu w-80 bg-base-100'>
          {menuItems?.map((m) => (
            <li key={m.name}>
              {m.internalURL && (
                <Link href={m.internalURL}>
                  <a
                    tabIndex={0}
                    role='link'
                    onKeyDown={() => {
                      if (checkboxRef.current) {
                        checkboxRef.current.checked = false
                      }
                    }}
                    onClick={() => {
                      if (checkboxRef.current) {
                        checkboxRef.current.checked = false
                      }
                    }}
                    className='rounded-btn'
                  >
                    {m.name}
                  </a>
                </Link>
              )}
              {m.externalURL && (
                <a
                  onClick={() => {
                    if (checkboxRef.current) {
                      checkboxRef.current.checked = false
                    }
                  }}
                  href={m.externalURL}
                  className='rounded-btn'
                >
                  {m.name}
                </a>
              )}
            </li>
          ))}
          <li>
            <div className='m-1'>
              {!user && (
                <Link href='/api/auth/login'>
                  <a className='btn btn-primary'>
                    Login
                    <BiLogIn size={20} className='mx-2' />
                  </a>
                </Link>
              )}

              {user && (
                <>
                  <DropDown
                    className='static bg-transparent'
                    width={50}
                    selectedId={null}
                    onSelect={() => {
                      /**/
                    }}
                    label={
                      <div className='avatar'>
                        <div className='w-10 h-10 m-1 rounded-full'>
                          {user?.picture ? (
                            <img alt={user.name || undefined} src={user?.picture} />
                          ) : (
                            <BsPersonFill />
                          )}
                        </div>
                      </div>
                    }
                    items={[
                      {
                        id: '1',
                        value: (
                          <Link href='/api/auth/logout'>
                            <a className='justify-center btn btn-ghost'>
                              <BiLogOut size={20} className='mx-2' />
                            </a>
                          </Link>
                        ),
                      },
                    ]}
                  />
                </>
              )}
            </div>
          </li>
        </ul>
      </div>
    </div>
  )
}
