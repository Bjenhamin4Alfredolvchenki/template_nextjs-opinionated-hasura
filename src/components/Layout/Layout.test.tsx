import React from 'react'
import renderer from 'react-test-renderer'
import { Layout } from './Layout'

test('renders correctly', () => {
  const tree = renderer.create(<Layout>Some Text</Layout>).toJSON()
  expect(tree).toMatchSnapshot()
})
