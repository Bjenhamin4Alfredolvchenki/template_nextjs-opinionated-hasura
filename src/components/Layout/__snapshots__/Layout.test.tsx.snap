// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="h-screen bg-base-100 drawer text-base-content"
>
  <input
    className="drawer-toggle"
    id="menu-drawer"
    type="checkbox"
  />
  <div
    className="flex flex-col drawer-content"
  >
    <div
      className="w-full navbar bg-base-300"
    >
      <div
        className="flex-none lg:hidden"
      >
        <label
          className="btn btn-square btn-ghost"
          htmlFor="menu-drawer"
        >
          <svg
            className="inline-block w-6 h-6 stroke-current"
            fill="none"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6h16M4 12h16M4 18h16"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth={2}
            />
          </svg>
        </label>
      </div>
      <div
        className="flex-none"
      >
        <div
          className="hidden lg:inline-block"
        >
          <ul
            className="space-x-2 menu horizontal"
          >
            <li>
              <div
                className="z-30 m-1"
              >
                <select
                  className="w-full max-w-xs select select-bordered"
                  onChange={[Function]}
                >
                  <option
                    disabled={true}
                  >
                    theme
                  </option>
                  <option
                    value="light"
                  >
                    Light
                  </option>
                  <option
                    value="dark"
                  >
                    Dark
                  </option>
                  <option
                    value="cupcake"
                  >
                    Cup Cake
                  </option>
                  <option
                    value="bumblebee"
                  >
                    Bumble Bee
                  </option>
                  <option
                    value="emerald"
                  >
                    Emerald
                  </option>
                  <option
                    value="corporate"
                  >
                    Corporate
                  </option>
                  <option
                    value="synthwave"
                  >
                    Synth Wave
                  </option>
                  <option
                    value="retro"
                  >
                    Retro
                  </option>
                  <option
                    value="cyberpunk"
                  >
                    Cyber-Punk
                  </option>
                  <option
                    value="valentine"
                  >
                    Valentine
                  </option>
                  <option
                    value="halloween"
                  >
                    Halloween
                  </option>
                  <option
                    value="garden"
                  >
                    Garden
                  </option>
                  <option
                    value="forest"
                  >
                    Forest
                  </option>
                  <option
                    value="aqua"
                  >
                    Aqua
                  </option>
                  <option
                    value="lofi"
                  >
                    Lofi
                  </option>
                  <option
                    value="pastel"
                  >
                    Pastel
                  </option>
                  <option
                    value="fantasy"
                  >
                    Fantasy
                  </option>
                  <option
                    value="wireframe"
                  >
                    Wire Frame
                  </option>
                  <option
                    value="black"
                  >
                    Black
                  </option>
                  <option
                    value="luxury"
                  >
                    Luxury
                  </option>
                  <option
                    value="dracula"
                  >
                    Dracula
                  </option>
                </select>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      className="w-full p-4 md:px-6 md:container md:mx-auto"
    >
      Some Text
    </div>
  </div>
  <div
    className="drawer-side"
  >
    <label
      className="drawer-overlay"
      htmlFor="menu-drawer"
    />
    <ul
      className="p-4 overflow-y-auto menu w-80 bg-base-100"
    />
  </div>
</div>
`;
